test/controllers/categories_controller_test.rb

```diff
@@ -0,0 +1,41 @@
+require "test_helper"
+
+class CategoriesControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    @category = Category.create(name: "Sports")
+    @admin_user = User.create(username: "vhodder", email: "vhodder@email.com",
+                              password: "password", admin: true)
+  end
+
+  test "should get index" do
+    get categories_url
+    assert_response :success
+  end
+
+  test "should get new" do
+    sign_in_as(@admin_user)
+    get new_category_url
+    assert_response :success
+  end
+
+  test "should create categroy" do
+    sign_in_as(@admin_user)
+    assert_difference('Category.count', 1) do
+      post categories_url, params: { category: { name: "Travel" } }
+    end
+    assert_redirected_to category_url(Category.last)
+  end
+
+  test "should not create categroy if not admin" do
+    assert_no_difference('Category.count') do
+      post categories_url, params: { category: { name: "Travel" } }
+    end
+    assert_redirected_to categories_url
+  end
+
+  test "should show category" do
```

Comment:
hier werden ja auch die Artikel zur Kategorie angezeigt und es wäre vielleicht eine gute Übung, das hier auch zu testen.

Reply:
Done.

Reply:
Wo wird das getestet?

Reply:
oh, ich glaube ich habe was anders gemacht und dacht, dass hat dein Bitte geantwortet :laughing: 
(Ich habe ein check geaddet, um das redirect zu testen in `test "should create category"` )

Ich kümmere mich darum.

Reply:
Also... ich kann sehr einfach testen, ob die Seite der Titel "Categories" hat mit: `assert_match "Categories", response.body`

Was ich jetzt bearbeite ist wie man checken kann, dass die Artiklen auch gelistet sind. `list_categories_test` macht so was ähnliches. Im Kurs haben wir das gemacht, indem wir geprüft haben, ob die Links zu den einzelnen Kategorien auf der Seite sind. In diesem Fall könnten wir prüfen, ob die Links zu den einzelnen Artikeln aufgelistet sind.

Ich habe es so ausprobiert:
1. ein @article instance in setup definieren, so : `@article = Article.create(title: "A Nice Title")`
Dann in `should show category` test : `assert_select "a[href=?]",  article_path(@article), text: @article.title`

Dies führt zu folgender Meldung:
```
ActionController::UrlGenerationError: No route matches {:action=>"show", :controller=>"articles", :id=>nil}, missing required keys: [:id]
```

Na gut, dann, option 2:
Ein @article instance in setup definieren, so : `@article = Article.create(id: 1, title: "A Nice Title")`

==> Yay, new error!
```
Expected at least 1 element matching "a[href="/articles/1"]", found 0..
Expected 0 to be >= 1.
```
Es scheint, dass es keine gute Idee ist, die id zu hardcoden.

Kennst du eine bessere Vorgehensweise, um dies zu tun/testen?

Reply:
Es reicht nicht, den Artikel anzulegen, Du musst ihm auch noch die Kategorie zuweisen, damit er hier erscheint...

Reply:
By the way: Ich habe das Test zu `"should show articles listing in category page"` umbenannt - für mich, ist es dann klarer was es so macht.

Reply:
Kann man das in das setup machen, so wie das hier?:
```
    @category = Category.create(id: 1, name: "Sports")
    @admin_user = User.create(username: "vhodder", email: "vhodder@email.com",
                              password: "password", admin: true)
    @article = Article.create(id: 1, title: "A Nice Title", category_ids: [@category.id])
```
Das funktioniert nicht, aber ich möchte wissen, ob ich in die richtige Richtung wandere...